File: mobile_app/lib/models/todo.dart
class Todo {
  final int? id;
  final String title;
  final bool completed;
  final String priority;
  final DateTime? createdAt;

  Todo({
    this.id,
    required this.title,
    this.completed = false,
    this.priority = 'MEDIUM',
    this.createdAt,
  });

  factory Todo.fromJson(Map<String, dynamic> json) {
    return Todo(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id']?.toString() ?? ''),
      title: json['title'] ?? '',
      completed: json['completed'] ?? false,
      priority: json['priority'] ?? 'MEDIUM',
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'title': title,
      'completed': completed,
      'priority': priority,
    };
    if (id != null) map['id'] = id;
    return map;
  }

  Todo copyWith({
    int? id,
    String? title,
    bool? completed,
    String? priority,
    DateTime? createdAt,
  }) {
    return Todo(
      id: id ?? this.id,
      title: title ?? this.title,
      completed: completed ?? this.completed,
      priority: priority ?? this.priority,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: backend/src/main/java/com/example/todoapp/model/Todo.java
package com.example.todoapp.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * Todo Entity representing a single task item.
 * Mapped to the 'todos' table in the H2 database.
 */
@Entity
@Table(name = "todos")
public class Todo {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    private boolean completed = false;

    @Column(nullable = false)
    private String priority = "MEDIUM"; // "HIGH", "MEDIUM", "LOW"

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    // Default constructor required by JPA
    public Todo() {
    }

    // Parameterized constructor for convenience
    public Todo(String title, String priority) {
        this.title = title;
        this.priority = priority != null ? priority : "MEDIUM";
        this.completed = false;
    }

    public Todo(Long id, String title, boolean completed, String priority) {
        this.id = id;
        this.title = title;
        this.completed = completed;
        this.priority = priority != null ? priority : "MEDIUM";
    }

    // Automatically set createdAt timestamp before persisting to database
    @PrePersist
    protected void onCreate() {
        if (this.createdAt == null) {
            this.createdAt = LocalDateTime.now();
        }
        if (this.priority == null || this.priority.trim().isEmpty()) {
            this.priority = "MEDIUM";
        }
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isCompleted() {
        return completed;
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
    }

    public String getPriority() {
        return priority;
    }

    public void setPriority(String priority) {
        this.priority = priority;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    @Override
    public String toString() {
        return "Todo{" +
                "id=" + id +
                ", title='" + title + '\'' +
                ", completed=" + completed +
                ", priority='" + priority + '\'' +
                ", createdAt=" + createdAt +
                '}';
    }
}
